keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+This KEP proposes implementing both the [Composition KEP] and the [Catalog KEP] as a prerequisite for plugin GA in Kustomize. Those two KEPs solve key discovery, distribution and UX problems described in the goals section.
+
+### Plugin execution restriction
+
+The flags used to restrict plugin execution will be replaced with a new set that reflects the GA plugin feature set and is consistent between standalone and embedded Kustomize.
+
+This KEP proposes using the concept of a "trusted plugin catalog" from the [Catalog KEP] as a convenient, concise way for end users to express their intent to execute a constrained set of plugins. With Catalog, the list of plugins being authorized is explicit and reviewable.
+
+**Plugins from a user-specified trusted catalog**
+
+When using plugins with a trusted catalog, only one flag will be required, regardless of the runtime type of the provider: `--trusted-plugin-catalog=""` .
+
+This new repeatable flag will be introduced by the [Catalog KEP]. When that flag is present, plugins in the referenced catalogs will be executed without the need for any additional flags. The integrity of the plugin provider will be validated against the Catalog entry before execution. This all applies whether the providers are defined explicitly in the Kustomization (and the explicit entry matches the catalog) or looked up dynamically in the catalog.
+
+If a containerized plugin needs network or disk access, its catalog entry MUST specify it. Redundant inline provider config cannot be used to grant it. When the catalog does specify it, no additional flags or config fields will be required to grant it the required access.
+
+**Plugins from a compiled-in trusted catalog**
+
+Optionally, it should be possible for a user to compile a bespoke distribution of Kustomize that embeds a trusted catalog, such that the plugins it references can be used without any additional flags on the end user's part. The target persona for this scenario is a platform maintainer at a large organization that distributes its own Kustomize internally. This approach could also be used in the future to extract Kustomize built-ins to plugins to reduce compiled in dependencies while minimizing UX impact.
+
+**Plugins NOT in a trusted catalog**
+
+*Recommended Option: no such thing*
+
+Trusted catalogs are always required to use plugins; uncatalogued plugins will not be executed. Kustomizations that require ad-hoc plugins must be accompanied by a `Catalog` that the end user can reference locally, e.g. `kustomize build dir/ --trusted-plugin-catalog=dir/catalog.yaml`. There will be no additional plugin-related flags. The end user will be expected to inspect this catalog before trusting it, which must always be done explicitly on the command line.
```

Comment:
Per previous comment I think this introduces additional burden on the plugin developer.

Reply:
It does, but it's a single command to run, and there are tremendous benefits to the end user as I see it, in terms of distribution and ease of reviewing what is being trusted. Those advantage apply even when the plugin is only used in one Kustomization, since in most cases use happens across multiple environments. IMO it's better to impose a minimal burden on the plugin developer than to impose a higher one on every end user / in every invocation context.